crates/wasmparser/src/binary_reader.rs

```diff
@@ -2507,3 +2258,552 @@ impl fmt::Debug for BrTable<'_> {
+    fn visit_else(&mut self, _offset: usize) -> Self::Output { Operator::Else }
+    fn visit_try(&mut self, _offset: usize, ty: BlockType) -> Self::Output { Operator::Try { ty } }
+    fn visit_catch(&mut self, _offset: usize, index: u32) -> Self::Output { Operator::Catch { index } }
+    fn visit_throw(&mut self, _offset: usize, index: u32) -> Self::Output { Operator::Throw { index } }
+    fn visit_rethrow(&mut self, _offset: usize, relative_depth: u32) -> Self::Output { Operator::Rethrow { relative_depth } }
+    fn visit_delegate(&mut self, _offset: usize, relative_depth: u32) -> Self::Output { Operator::Delegate { relative_depth } }
+    fn visit_catch_all(&mut self, _offset: usize) -> Self::Output { Operator::CatchAll }
+    fn visit_end(&mut self, _offset: usize) -> Self::Output { Operator::End }
+    fn visit_br(&mut self, _offset: usize, relative_depth: u32) -> Self::Output { Operator::Br { relative_depth } }
+    fn visit_br_if(&mut self, _offset: usize, relative_depth: u32) -> Self::Output { Operator::BrIf { relative_depth } }
+    fn visit_br_table(&mut self, _offset: usize, table: &BrTable<'a>) -> Self::Output { Operator::BrTable { table: table.clone() } }
+    fn visit_return(&mut self, _offset: usize) -> Self::Output { Operator::Return }
+    fn visit_call(&mut self, _offset: usize, function_index: u32) -> Self::Output { Operator::Call { function_index } }
+    fn visit_return_call(&mut self, _offset: usize, function_index: u32) -> Self::Output { Operator::ReturnCall { function_index } }
+    fn visit_call_indirect(&mut self, _offset: usize, index: u32, table_index: u32, table_byte: u8) -> Self::Output { Operator::CallIndirect { index, table_index, table_byte } }
+    fn visit_return_call_indirect(&mut self, _offset: usize, index: u32, table_index: u32) -> Self::Output { Operator::ReturnCallIndirect { index, table_index } }
+    fn visit_drop(&mut self, _offset: usize) -> Self::Output { Operator::Drop }
+    fn visit_select(&mut self, _offset: usize) -> Self::Output { Operator::Select }
+    fn visit_typed_select(&mut self, _offset: usize, ty: ValType) -> Self::Output { Operator::TypedSelect { ty } }
+    fn visit_local_get(&mut self, _offset: usize, local_index: u32) -> Self::Output { Operator::LocalGet { local_index } }
+    fn visit_local_set(&mut self, _offset: usize, local_index: u32) -> Self::Output { Operator::LocalSet { local_index } }
+    fn visit_local_tee(&mut self, _offset: usize, local_index: u32) -> Self::Output { Operator::LocalTee { local_index } }
+    fn visit_global_get(&mut self, _offset: usize, global_index: u32) -> Self::Output { Operator::GlobalGet { global_index } }
+    fn visit_global_set(&mut self, _offset: usize, global_index: u32) -> Self::Output { Operator::GlobalSet { global_index } }
+    fn visit_i32_load(&mut self, _offset: usize, memarg: MemoryImmediate) -> Self::Output { Operator::I32Load { memarg } }
```

Comment:
One perf-related thing I might recommend looking into if you're interested is to pass larger-than-pointer-sized-values, like `MemoryImmediate` and some of the vector-related ones by pointer instead of by value. That may make some operations here more efficient (not this specific impl per se but the other impls)

Reply:
Cool! Thanks for the info. Gonna try that out and post results.

Reply:
Cool! Thanks for the info. Gonna try that out and post results.

Reply:
I implemented it and hopefully did not forget any of the variants but benchmarks indicated no real performance difference.

Reply:
I tried various tricks for this and was unable to get more than a 1% perf improvement.

My rough thinking is that passing `MemoryImmediate` and `V128Const` and such by value require allocating extra stack space in the visit function or otherwise more memory copies since things are passed by value. My thinking was that visiting an operator would pass in `&mut OperatorImmediates` which is allocated in some caller-specified location which means that, theoretically, `visit_operator` can work with a smaller stack frame. 

In practice it doesn't seem to help much though so I don't think it's worth pursuing at this time.